src/pudl/extract/excel.py

```diff
@@ -0,0 +1,262 @@
+
+    def get_all_columns(self, page):
+        """Returns list of all pudl (standardized) columns for a given page (across all years)."""
+        return sorted(self._column_map[page].columns)
+
+    def get_all_pages(self):
+        """Returns list of all known pages."""
+        return sorted(self._column_map.keys())
+
+    @staticmethod
+    def _load_csv(package, filename):
+        """Load metadata from a filename that is found in a package."""
+        return pd.read_csv(importlib.resources.open_text(package, filename),
+                           index_col=0, comment='#')
+
+
+class GenericExtractor(object):
+    """Contains logic for extracting panda.DataFrames from excel spreadsheets.
+
+    This class implements the generic dataset agnostic logic to load data
+    from excel spreadsheet simply by using excel Metadata for given dataset.
+
+    It is expected that individual datasets wil subclass this code and add
+    custom business logic by overriding necessary methods.
+
+    When implementing custom business logic, the following should be modified:
+
+    1. DATASET class attribute controls which excel metadata should be loaded.
+
+    2. BLACKLISTED_PAGES class attribute specifies which pages should not
+    be loaded from the underlying excel files even if the metadata is
+    available. This can be used for experimental/new code that should not be
+    run yet.
+
+    3. file_basename_glob() tells us what is the basename of the excel file
+    that contains the data for a given (year, page).
+
+    4. dtypes() should return dict with {column_name: pandas_datatype} if you
+    need to specify which datatypes should be uded upon loading.
+
+    5. If data cleanup is necessary, you can apply custom logic by overriding
+    one of the following functions (they all return the modified dataframe):
+    - process_raw() is applied right after loading the excel DataFrame
+    from the disk.
+    - process_renamed() is applied after input columns were renamed to
+    standardized pudl columns.
+    - process_final_page() is applied when data from all available years
+    is merged into single DataFrame for a given page.
+    """
+
+    METADATA = None
+    """Instance of metadata object to use with this extractor."""
+
+    BLACKLISTED_PAGES = []
+    """List of supported pages that should not be extracted."""
+
+    def __init__(self, data_dir, metadata=None):
+        """Create new extractor object and load metadata.
+
+        Args:
+            data_dir: Path to the data_dir to use when loading excel
+              files from disk (passed to datastore).
+        """
+        self._data_dir = data_dir
+        if not self.METADATA:
+            raise NotImplementedError('self.METADATA must be set.')
+        self._metadata = self.METADATA
+        self._dataset_name = self._metadata.get_dataset_name()
+        self._file_cache = {}
+
+    def process_raw(self, df, year, page):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Method could be a function](https://app.codacy.com/manual/zaneselvans/pudl/pullRequest?prid=5191218)